compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/writer/BIRInstructionWriter.java

```diff
@@ -0,0 +1,606 @@
+        writePosition(birNewStructure.pos);
+        buf.writeByte(birNewStructure.kind.getValue());
+        birNewStructure.rhsOp.accept(this);
+        birNewStructure.lhsOp.accept(this);
+    }
+
+    public void visit(BIRNonTerminator.NewInstance newInstance) {
+        writePosition(newInstance.pos);
+        buf.writeByte(newInstance.kind.getValue());
+        buf.writeBoolean(newInstance.isExternalDef);
+        if (newInstance.isExternalDef) {
+            assert newInstance.externalPackageId != null;
+            buf.writeInt(addPkgCPEntry(newInstance.externalPackageId));
+            buf.writeInt(addStringCPEntry(newInstance.objectName));
+        } else {
+            buf.writeInt(newInstance.def.index);
+        }
+        newInstance.lhsOp.accept(this);
+    }
+
+    public void visit(NewArray birNewArray) {
+        writePosition(birNewArray.pos);
+        buf.writeByte(birNewArray.kind.getValue());
+        writeType(birNewArray.type);
+        birNewArray.lhsOp.accept(this);
+        birNewArray.sizeOp.accept(this);
+    }
+
+    public void visit(BIRNonTerminator.FieldAccess birFieldAccess) {
+        writePosition(birFieldAccess.pos);
+        buf.writeByte(birFieldAccess.kind.getValue());
+        if (birFieldAccess.kind == InstructionKind.MAP_LOAD || birFieldAccess.kind == InstructionKind.ARRAY_LOAD) {
+            buf.writeBoolean(birFieldAccess.optionalFieldAccess);
+            buf.writeBoolean(birFieldAccess.fillingRead);
+        }
+        birFieldAccess.lhsOp.accept(this);
+        birFieldAccess.keyOp.accept(this);
+        birFieldAccess.rhsOp.accept(this);
+    }
+
+    public void visit(BIRNonTerminator.TypeCast birTypeCast) {
+        writePosition(birTypeCast.pos);
+        buf.writeByte(birTypeCast.kind.getValue());
+        birTypeCast.lhsOp.accept(this);
+        birTypeCast.rhsOp.accept(this);
+        writeType(birTypeCast.type);
+        buf.writeBoolean(birTypeCast.checkTypes);
+    }
+
+    public void visit(BIRNonTerminator.IsLike birIsLike) {
+        writePosition(birIsLike.pos);
+        buf.writeByte(birIsLike.kind.getValue());
+        writeType(birIsLike.type);
+        birIsLike.lhsOp.accept(this);
+        birIsLike.rhsOp.accept(this);
+    }
+
+    public void visit(BIRNonTerminator.TypeTest birTypeTest) {
+        writePosition(birTypeTest.pos);
+        buf.writeByte(birTypeTest.kind.getValue());
+        writeType(birTypeTest.type);
+        birTypeTest.lhsOp.accept(this);
+        birTypeTest.rhsOp.accept(this);
+    }
+
+    // Operands
+    public void visit(BIROperand birOperand) {
+        if (birOperand.variableDcl.ignoreVariable) {
+            buf.writeBoolean(true);
+            writeType(birOperand.variableDcl.type);
+            return;
+        }
+
+        buf.writeBoolean(false);
+        buf.writeByte(birOperand.variableDcl.kind.getValue());
+        buf.writeByte(birOperand.variableDcl.scope.getValue());
+
+        // TODO use the integer index of the variable.
```

Comment:
Shall we create issues for Todos so they will be fixed later?

Reply:
refer https://github.com/ballerina-platform/ballerina-lang/issues/25440